woocommerce/woocommerce-ios
# [Product Multi-Selection ] Add and remove products and variations

Closes: #9196 

Apologies for the large PR, most of the changes come from adding a `setUp()` method to Unit Tests, and tests cases for single-selection vs multi-selection.

## Description
This PR implements adding  and removing multiple items (Products or Product Variations) to an Order when tapping on the "X Products Selected" button, as well as keeping selected/unselected products in synchronisation with the state of the Order:

- Selecting items that are not yet in the Order, and tapping  "X Products Selected" will add them.
- Unselecting items that are already in the Order, and tapping  "X Products Selected" will remove them.
- Removing items from the Order creation/editing view will update the selections state if we open the product selector.
- Opening existing Orders with items will show selected items.

![Simulator Screen Recording - iPhone Xs - 2023-03-20 at 11 08 21](https://user-images.githubusercontent.com/3812076/226245082-ff1c23c3-703e-4507-baa0-8fbb2033b2e9.gif)


## Changes
* Adds a `onMultipleSelectionCompleted` completion handler to `ProductSelectorViewModel`, which is be triggered when the `(X) products selected` button is tapped.
* This completion handler calls `syncOrderItems()` which deals with compiling all the items that need to be sync ( both additions and removals ) into a unique `[OrderSyncProductInput]` array. This is necessary as we need to trigger the call to the remote synchroniser only once for all changes, otherwise some changes may be ignored.
* `selectExistingProductsInOrderIfNeeded()` is called on `EditableOrderViewModel`'s initialization, this checks the latest Order sync, and is used so we can keep the selected state for different products when an Order already contains those.

## Testing instructions
- Go to Menu > Settings > Experimental features > and enable `Product Multi-Selection`
- Go to Orders > Tap `+` > Tap on `+ Add Products` > Add different products to an Order > Tap "X Products Selected" 
- See how the Order is updated with all products.
- Removing products either from the Order view or by tapping `+ Add Products`, deselecting them, and tapping on  "X Products Selected"  will update the Order by removing these products.

### Suggested testing flows:
- Add, and remove products
- Add, and remove product variations
- Remove items from the Order view by tapping on `-`
- Create a new Order and confirm that all selected/unselected products are correct.
- Edit an existing Order to confirm that all selected products are correct.
